File: src/main/java/sage/domain/service/GroupService.java
package sage.domain.service;

import java.util.Collection;
import java.util.Date;
import java.util.Set;

import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import sage.domain.commons.DomainRuntimeException;
import sage.domain.commons.IdCommons;
import sage.domain.repository.*;
import sage.entity.*;
import sage.transfer.GroupPreview;
import sage.transfer.UserLabel;
import sage.util.Colls;

@Service
@Transactional
public class GroupService {
  @Autowired
  private GroupRepository groupRepo;
  @Autowired
  private TopicPostRepository topicPostRepo;
  @Autowired
  private TagRepository tagRepo;
  @Autowired
  private UserRepository userRepo;
  @Autowired
  private BlogRepository blogRepo;

  public GroupPreview getGroupPreview(long id) {
    return new GroupPreview(groupRepo.nonNull(id));
  }

  public Collection<UserLabel> members(long groupId) {
    return UserLabel.listOf(groupRepo.nonNull(groupId).getMembers());
  }

  public GroupPreview create(long userId, String name, String introduction, Collection<Long> tagIds) {
    if (name == null || name.isEmpty()) {
      throw new DomainRuntimeException("Must enter a group name!");
    }
    if (introduction == null || introduction.isEmpty()) {
      throw new DomainRuntimeException("Must enter a group introduction!");
    }

    Set<Tag> tags = tagRepo.byIds(tagIds);
    Group group = groupRepo.save(new Group(name, introduction, tags, userRepo.load(userId), new Date()));
    return new GroupPreview(group);
  }

  public Group edit(long userId, long groupId,
                    String name, String introduction, Collection<Long> tagIds) {
    if (name == null || name.isEmpty()) {
      throw new DomainRuntimeException("Must enter a group name!");
    }
    if (introduction == null || introduction.isEmpty()) {
      throw new DomainRuntimeException("Must enter a group introduction!");
    }
    Group group = groupRepo.nonNull(groupId);
    if (group == null) {
      throw new DomainRuntimeException("Group[%d] does not exist!", groupId);
    }
    if (!group.getCreator().getId().equals(userId)) {
      throw new DomainRuntimeException("User[%d] is not the owner of Group[%d]", userId, groupId);
    }

    group.setName(name);
    group.setIntroduction(introduction);
    group.setTags(tagRepo.byIds(tagIds));
    groupRepo.update(group);

    return group;
  }

  public void join(long userId, long groupId) {
    User user = userRepo.load(userId);
    Group group = groupRepo.nonNull(groupId);
    if (group.getMembers().add(user)) {
      groupRepo.update(group);
    }
  }

  public void exit(long userId, long groupId) {
    User user = userRepo.load(userId);
    Group group = groupRepo.nonNull(groupId);
    if (group.getMembers().remove(user)) {
      groupRepo.update(group);
    }
  }

  public Collection<GroupPreview> byTags(Collection<Long> tagIds) {
    return Colls.map(groupRepo.byTags(tagRepo.byIds(tagIds)), GroupPreview::new);
  }

  public Collection<GroupPreview> all() {
    return Colls.map(groupRepo.all(), GroupPreview::new);
  }
}

File: src/main/kotlin/sage/service/GroupService.kt
package sage.service

import org.springframework.stereotype.Service
import sage.domain.commons.DomainException
import sage.entity.Group
import sage.entity.Tag
import sage.entity.User
import sage.transfer.GroupPreview
import sage.transfer.UserLabel

@Service
class GroupService {

  fun getGroupPreview(id: Long): GroupPreview {
    return GroupPreview(Group.get(id))
  }

  fun members(groupId: Long): Collection<UserLabel> {
    return Group.get(groupId).members.map { UserLabel(it) }
  }

  fun create(userId: Long, name: String, introduction: String, tagIds: Collection<Long>): GroupPreview {
    if (name.isEmpty()) {
      throw DomainException("Must enter a group name!")
    }
    if (introduction.isEmpty()) {
      throw DomainException("Must enter a group introduction!")
    }

    val tags = Tag.multiGet(tagIds)
    val group = Group(name, introduction, tags, User.ref(userId))
    group.save()
    return GroupPreview(group)
  }

  fun edit(userId: Long, groupId: Long,
           name: String, introduction: String, tagIds: Collection<Long>): Group {
    if (name.isEmpty()) {
      throw DomainException("Must enter a group name!")
    }
    if (introduction.isEmpty()) {
      throw DomainException("Must enter a group introduction!")
    }
    val group = Group.get(groupId)
    if (userId != group.creator?.id) {
      throw DomainException("User[%d] is not the owner of Group[%d]", userId, groupId)
    }
    group.name = name
    group.introduction = introduction
    group.tags = Tag.multiGet(tagIds)
    group.update()
    return group
  }

  fun join(userId: Long, groupId: Long) {
    val user = User.ref(userId)
    val group = Group.get(groupId)
    if (group.members.add(user)) {
      group.update()
    }
  }

  fun exit(userId: Long, groupId: Long) {
    val user = User.ref(userId)
    val group = Group.get(groupId)
    if (group.members.remove(user)) {
      group.update()
    }
  }

  fun byTags(tagIds: Collection<Long>) =
      Group.where().filterMany("tags").`in`("id", tagIds).findList().map { GroupPreview(it) }

  fun all() = Group.all().map { GroupPreview(it) }
}
